File: haskell/leetcode/dynamic_programming.hs
module DynamicProgramming where
-- TOPIC: Dynamic Programming LeetCode Problems | ghci: :load dynamic_programming.hs
import Data.Array (Array, array, (!), bounds, listArray)

-- TODO 1: Climbing Stairs — LC #70
-- climbStairs :: Int -> Int
-- climbStairs n = fibs !! n
--   where fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

-- TODO 2: Coin Change — LC #322
-- coinChange :: [Int] -> Int -> Int
-- coinChange coins amount = dp ! amount
--   where dp = array (0, amount) [(i, solve i) | i <- [0..amount]]
--         solve 0 = 0
--         solve i = minimum $ (-1) : [dp ! (i-c) + 1 | c <- coins, c <= i, dp ! (i-c) >= 0]

-- TODO 3: Longest Increasing Subsequence — LC #300
-- lengthOfLIS :: [Int] -> Int
-- -- O(n²): dp[i] = 1 + max (dp[j] | j < i, nums[j] < nums[i])

-- TODO 4: Longest Common Subsequence — LC #1143
-- longestCommonSubsequence :: String -> String -> Int
-- lcs s t = dp ! (m, n)
--   where m = length s; n = length t
--         sa = listArray (1,m) s; ta = listArray (1,n) t
--         dp = array ((0,0),(m,n)) [((i,j), go i j) | i<-[0..m], j<-[0..n]]
--         go 0 _ = 0; go _ 0 = 0
--         go i j = if sa!i == ta!j then dp!(i-1,j-1)+1 else max (dp!(i-1,j)) (dp!(i,j-1))

-- TODO 5: 0/1 Knapsack (classic)
-- knapsack :: [Int] -> [Int] -> Int -> Int
-- -- dp[i][w] = max value with first i items, capacity w

-- TODO 6: Decode Ways — LC #91
-- numDecodings :: String -> Int
-- -- dp[i] = ways to decode s[0..i-1]

-- TODO 7: Word Break — LC #139
-- wordBreak :: String -> [String] -> Bool
-- -- dp[i] = can s[0..i-1] be segmented

File: kotlin/leetcode/dynamic_programming.kt
package leetcode
// TOPIC: Dynamic Programming LeetCode Problems | kotlinc dynamic_programming.kt -include-runtime -d out.jar && java -jar out.jar

// TODO 1: Climbing Stairs — LC #70
//   fun climbStairs(n: Int): Int
//   - dp[i] = dp[i-1] + dp[i-2]; optimize to two variables

// TODO 2: Coin Change — LC #322
//   fun coinChange(coins: IntArray, amount: Int): Int
//   - dp[i] = min coins for amount i; Int.MAX_VALUE for unreachable

// TODO 3: Longest Increasing Subsequence — LC #300
//   fun lengthOfLIS(nums: IntArray): Int
//   - O(n log n) with patience sorting using binarySearch

// TODO 4: Longest Common Subsequence — LC #1143
//   fun longestCommonSubsequence(text1: String, text2: String): Int
//   - 2D DP table: dp[i][j] = LCS of text1[0..i] and text2[0..j]

// TODO 5: 0/1 Knapsack (classic)
//   fun knapsack(weights: IntArray, values: IntArray, capacity: Int): Int
//   - dp[i][w] = max value using first i items with capacity w

// TODO 6: Decode Ways — LC #91
//   fun numDecodings(s: String): Int
//   - dp[i] = ways to decode s[0..i-1]

// TODO 7: Word Break — LC #139
//   fun wordBreak(s: String, wordDict: List<String>): Boolean
//   - dp[i] = can s[0..i-1] be segmented

// TODO 8: House Robber — LC #198
//   fun rob(nums: IntArray): Int
//   - dp[i] = max(dp[i-1], dp[i-2] + nums[i])

fun main() {}
